### DISCUSSION TITLE
Can use of `internal` instead of `public` decrease build times?

### DISCUSSION THREAD

**Discussion by MartyIX** [dotnet/msbuild#10120] [Q&A]
Perhaps this is a silly question but I'll try anyway. I have a .NET solution in .NET 8 with say 20 projects at this point. Now there is a shared project `SharedProject` and there are `class`es, and `struct`s in this project that are `public` but they can be turned into `internal` (AFAIK).

My question is: If I succeed in changing some types to `internal`, can this change lead to faster solution rebuilds in some cases? The idea being if an `internal` type is changed, then perhaps msbuild can accelerate builds in more scenarios or use some heuristics.

Please don't tell me that the types should be `internal` in the first place. That's not what this question is about, I want to get better grasp how msbuilds behaves[^1].

[^1]: Yes, I should probably measure it using a binlog.

**Comment by KalleOlaviNiemitalo** [comment]
Yes, changing things from public to internal can speed up incremental builds, if the projects produce reference assemblies and some referencing projects use those reference assemblies.  See the ProduceReferenceAssembly and CompileUsingReferenceAssemblies properties.

If anything in the shared project (`*.shproj`) changes, then the projects that reference the shared project have to be rebuilt.  But any projects that reference those projects can be considered up to date if the reference assemblies did not change.

**Reply by MartyIX** [reply]
That's interesting. Thank you for the comment.

So for normal assemblies there is no speedup, right?
